[juju] Add juju integration to sos collect.

chanchiwai-ray opened the pull request:
Add a new cluster profile and transport called "juju" for `sos collect`. Both the profile and transport are intended to be used on juju managed environments which assumes that `juju` is installed on the machine where `sos collect` is called, and that the juju user has superuser privilege to the current controller.

When using the "juju" cluster profile, the sos reports will be collected from all the applications within the current model by default. If necessary, one can filter the nodes by models / applications / units / machines with cluster options. For example `-c "juju.models=sos" -c "juju.apps=a,b,c"`. Moreover, transport will also be dynamically changed to "juju" when cluster type is juju.

If not using "juju" cluster profile, one can still choose to use the "juju" transport by specifying --transport option. However, not that the --nodes option will be expected to be a comma separated machine IDs, **not** IP addr, since `juju ssh` identifies the ssh target by machine ID. For example, `sos collect --nodes 0,1,2`.
Closes: #2668

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [ ] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [ ] Is the subject and message clear and concise?
- [ ] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [ ] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [ ] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?

chanchiwai-ray committed:
[juju] Add juju integration to sos collect.

Add a new cluster profile and transport called "juju" for `sos collect`.
Both the profile and transport are intended to be used on juju managed
environments which assumes that `juju` is installed on the machine where
`sos collect` is called, and that the juju user has superuser privilege
to the current controller.

When using the "juju" cluster profile, the sos reports will be collected
from all the applications within the current model by default. If
necessary, one can filter the nodes by models / applications / units /
machines with cluster options. For example `-c "juju.models=sos" -c
"juju.apps=a,b,c"`. Moreover, transport will also be dynamically changed
to "juju" when cluster type is juju.

If not using "juju" cluster profile, one can still choose to use the
"juju" transport by specifying --transport option. However, not that the
--nodes option will be expected to be a comma separated machine IDs ,
**not** IP addr, since `juju ssh` identifies the ssh target by machine
ID. For example, `sos collect --nodes 0,1,2`

Closes: #2668

Signed-off-by: Chi Wai Chan <chiwai.chan@canonical.com>

esunar commented on sos/collector/clusters/juju.py:
```diff
@@ -0,0 +1,115 @@
+    cmd = "juju"
+    cluster_name = "Juju Managed Clouds"
+    option_list = [
+        ("apps", "", "Filter node list by apps (comma separated)."),
```
should we support regex? It is not much work and it can be useful. for example for all ceph related apps `.*ceph.*`

esunar commented on sos/collector/clusters/juju.py:
```diff
@@ -0,0 +1,109 @@
+
+    Example:
+
+    sos collect --cluster-type juju -c "juju.models=sos" -c "juju.apps=a,b,c"
```
Should we change something like `sos collect --cluster-type juju -c "juju.models=lma juju.apps=graylog" -c "juju.models=openstack juju.apps=glance"` to give the ability to collect from multiple models at the same time? Or is it overkill?

esunar replied:
we should not

esunar commented on sos/collector/clusters/juju.py:
```diff
@@ -0,0 +1,109 @@
+    It"s assumed that `juju` is installed on the machine where `sos` is called,
+    and that the juju user has superuser privilege to the current controller.
+
+    By default, the sos reports will be collected from all the applications in
```
Meybe we need to change the default behaviour.[ Because it may cause trouble in large clouds](https://github.com/chanchiwai-ray/sos/pull/1#pullrequestreview-1370512330).

esunar commented on sos/collector/exceptions.py:
```diff
@@ -124,5 +132,6 @@ def __init__(self):
     'UnsupportedHostException',
-    'InvalidTransportException'
+    'InvalidTransportException',
+    'JujuNotInstalledException'
```
can you add trailing comma

esunar commented:
We should add more debug code. Especially to transport.

Commit:
Cleanup
Added regex support for filtering apps,units
Added subordinates for filtering
Some refactoring

Commit:
Fix for parameter duplication

Commit:
refactor

jneo8 commented on sos/collector/transports/juju.py:
```diff
@@ -0,0 +1,81 @@
+            raise
+        return True
+
+    def _connect(self, password=''):
```
**issue:** 

Why we need password here?

jneo8 commented on sos/collector/transports/juju.py:
```diff
@@ -0,0 +1,81 @@
+
+    @property
+    def connected(self):
+        return self._connected
```
**nitpick:** 

`self._connected` seems can be removed.

functions `_connect` and `connected` seems can be merged.

chanchiwai-ray commented:
Superseded by #3192
